developer-docs-site/docs/guides/sign-a-transaction.md

```diff
@@ -0,0 +1,311 @@
+    "max_gas_amount": 2000,
+    "gas_unit_price": 1,
+    // Unix timestamp, in seconds + 10 minutes
+    "expiration_timestamp_secs": Math.floor(Date.now() / 1000) + 600,
+    "payload": payload,
+    "chain_id": 3
+  };
+}
+
+```
+
+## Step 2) Create a signing message
+
+1. [SHA3_256][sha3] hash bytes of string "APTOS::RawTransaction"
+2. Bytes of BCS serialized RawTransaction
+3. Concat the prefix and BCS bytes
+
+```typescript
+
+class BCS {
+  static uleb128_encoded_bytes(num: number): Buffer {
+    if (num === 0)
+        return Buffer.from([0]);
+
+    const arr = [];
+    let len = 0;
+
+    while (num > 0) {
+        arr[len] = num & 0x7F;
+        if (num >>= 7) arr[len] |= 0x80;
+        len++;
+    }
+
+    return Buffer.from(arr);
+  }
+
+  static serialize_size(num: number): Buffer {
+    return BCS.uleb128_encoded_bytes(num);
+  }
+
+  static serialize_string(str: string): Buffer {
+    return BCS.serialize_bytes(Uint8Array.from(Buffer.from(str, 'ascii')));
+  }
+
+  static serialize_bytes(bytes: Uint8Array): Buffer {
+    const buf = BCS.serialize_size(bytes.length);
+    return Buffer.concat([buf, Buffer.from(bytes)]);
+  }
+
+  static serialize_uint64(num: bigint): Buffer {
+    const arr = new ArrayBuffer(8);
+    const view = new DataView(arr);
+    view.setBigUint64(0, num, true); // little endian
+    return Buffer.from(arr);
+  }
+
+  static serialize_uint8(num: bigint): Buffer {
+    const arr = new ArrayBuffer(1);
+    const view = new DataView(arr);
+    view.setInt8(0, num);
+    return Buffer.from(arr);
+  }
+}
+
+function hashPrefix(): Buffer {
+  let hash = SHA3.sha3_256.create();
+  // APTOS::RawTransaction
+  hash.update(`${HASH_PREFIX}RawTransaction`);
+  return Buffer.from(hash.arrayBuffer());
+}
+
+function bcsSerializeTransactionPayload(payload: TransactionPayload): Buffer {
+  return Buffer.concat([
+    BCS.serialize_string(payload.type),
+    BCS.serialize_string(payload.function),
+    ...payload.type_arguments.map(tyArg => BCS.serialize_string(tyArg)),
+    ...payload.arguments.map(arg => BCS.serialize_string(arg))
+  ]);
+}
+
+function bcsSerializeRawTransaction(txn: RawTransaction): Buffer {
+  return Buffer.concat([
+    BCS.serialize_string(txn.sender),
+    BCS.serialize_uint64(txn.sequence_number),
+    bcsSerializeTransactionPayload(txn.payload),
+    BCS.serialize_uint64(txn.max_gas_amount),
+    BCS.serialize_uint64(txn.gas_unit_price),
+    BCS.serialize_uint64(txn.expiration_timestamp_secs),
+    BCS.serialize_uint8(txn.chain_id),
+  ]);
+}
+
+// This will hash a raw transaction into bytes
+function hashRawTransaction(txn: RawTransaction): Buffer {
+  // Generate a hash prefix
+  const prefix = hashPrefix();
+
+  // Serialize txn with BCS
+  const bcsSerializedTxn = bcsSerializeRawTransaction(txn);
+
+  return Buffer.concat([prefix, bcsSerializedTxn]);
+}
+
+```
+
+## Step 3) Sign the signing message with your account private key
+
+Sign the hashed raw transaction bytes with [Ed25519][ed25519]
```

Comment:
start first with single signer and then leave space for the other variants.

Reply:
👍